Build process errors with latest protobuf update

Following #1660, fresh `git clone`s of `github.com/grpc/grpc.git` master (no prior system installation of `grpc` or `protobuf`) fail to build against GCC 4.8 on Linux (Ubuntu 14.04 and 15.04) with the following steps:

```
git clone https://github.com/grpc/grpc
cd grpc
git submodule update --init
make
```

The error is:
```
In file included from /usr/include/c++/4.8/unordered_map:47:0,
                 from ./google/protobuf/stubs/hash.h:44,
                 from ./google/protobuf/map.h:35,
                 from ./google/protobuf/struct.pb.h:29,
                 from google/protobuf/struct.pb.cc:5:
/usr/include/c++/4.8/bits/hashtable.h: In instantiation of 'std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__node_type* std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::_M_allocate_node(_Args&& ...) [with _Args = {const std::piecewise_construct_t&, std::tuple<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >&>, std::tuple<>}; _Key = std::basic_string<char>; _Value = std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>; _Alloc = google::protobuf::Map<std::basic_string<char>, google::protobuf::Value>::MapAllocator<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*> >; _ExtractKey = std::__detail::_Select1st; _Equal = std::equal_to<std::basic_string<char> >; _H1 = google::protobuf::hash<std::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; std::_Hashtable<_Key, _Value, _Alloc, _ExtractKey, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits>::__node_type = std::__detail::_Hash_node<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, true>]':
/usr/include/c++/4.8/bits/hashtable_policy.h:493:8:   required from 'std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::mapped_type& std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::operator[](const key_type&) [with _Key = std::basic_string<char>; _Pair = std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>; _Alloc = google::protobuf::Map<std::basic_string<char>, google::protobuf::Value>::MapAllocator<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*> >; _Equal = std::equal_to<std::basic_string<char> >; _H1 = google::protobuf::hash<std::basic_string<char> >; _H2 = std::__detail::_Mod_range_hashing; _Hash = std::__detail::_Default_ranged_hash; _RehashPolicy = std::__detail::_Prime_rehash_policy; _Traits = std::__detail::_Hashtable_traits<true, false, true>; std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::mapped_type = google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*; std::__detail::_Map_base<_Key, _Pair, _Alloc, std::__detail::_Select1st, _Equal, _H1, _H2, _Hash, _RehashPolicy, _Traits, true>::key_type = std::basic_string<char>]'
/usr/include/c++/4.8/bits/unordered_map.h:596:20:   required from 'std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::mapped_type& std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::operator[](const key_type&) [with _Key = std::basic_string<char>; _Tp = google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*; _Hash = google::protobuf::hash<std::basic_string<char> >; _Pred = std::equal_to<std::basic_string<char> >; _Alloc = google::protobuf::Map<std::basic_string<char>, google::protobuf::Value>::MapAllocator<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*> >; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::mapped_type = google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*; std::unordered_map<_Key, _Tp, _Hash, _Pred, _Alloc>::key_type = std::basic_string<char>]'
./google/protobuf/map.h:277:36:   required from 'T& google::protobuf::Map<Key, T>::operator[](const key_type&) [with Key = std::basic_string<char>; T = google::protobuf::Value; google::protobuf::Map<Key, T>::key_type = std::basic_string<char>]'
google/protobuf/struct.pb.cc:302:43:   required from here
/usr/include/c++/4.8/bits/hashtable.h:727:6: error: no matching function for call to 'google::protobuf::Map<std::basic_string<char>, google::protobuf::Value>::MapAllocator<std::__detail::_Hash_node<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, true> >::construct(std::_Hashtable<std::basic_string<char>, std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, google::protobuf::Map<std::basic_string<char>, google::protobuf::Value>::MapAllocator<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*> >, std::__detail::_Select1st, std::equal_to<std::basic_string<char> >, google::protobuf::hash<std::basic_string<char> >, std::__detail::_Mod_range_hashing, std::__detail::_Default_ranged_hash, std::__detail::_Prime_rehash_policy, std::__detail::_Hashtable_traits<true, false, true> >::__node_type*&, const std::piecewise_construct_t&, std::tuple<const std::basic_string<char, std::char_traits<char>, std::allocator<char> >&>, std::tuple<>)'
      _M_node_allocator().construct(__n, std::forward<_Args>(__args)...);
      ^
/usr/include/c++/4.8/bits/hashtable.h:727:6: note: candidate is:
In file included from ./google/protobuf/struct.pb.h:29:0,
                 from google/protobuf/struct.pb.cc:5:
./google/protobuf/map.h:170:10: note: void google::protobuf::Map<Key, T>::MapAllocator<U>::construct(google::protobuf::Map<Key, T>::MapAllocator<U>::pointer, google::protobuf::Map<Key, T>::MapAllocator<U>::const_reference) [with U = std::__detail::_Hash_node<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, true>; Key = std::basic_string<char>; T = google::protobuf::Value; google::protobuf::Map<Key, T>::MapAllocator<U>::pointer = std::__detail::_Hash_node<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, true>*; google::protobuf::Map<Key, T>::MapAllocator<U>::value_type = std::__detail::_Hash_node<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, true>; google::protobuf::Map<Key, T>::MapAllocator<U>::const_reference = const std::__detail::_Hash_node<std::pair<const std::basic_string<char>, google::protobuf::MapPair<std::basic_string<char>, google::protobuf::Value>*>, true>&]
     void construct(pointer p, const_reference t) { new (p) value_type(t); }
          ^
./google/protobuf/map.h:170:10: note:   candidate expects 2 arguments, 4 provided
```

---

Against GCC 4.9, the above works fine.

Performing the following series of steps builds everything properly against GCC 4.8:
```
git clone https://github.com/grpc/grpc
cd grpc
git submodule update --init
cd third_party/protobuf
./autogen.sh
./configure
make
sudo make install
cd ../..          # cd to grpc root
make
```

It is thus suspected that the `Makefile` for `grpc` is somehow polluting the configuration step for `protobuf` in a manner sensitive to GCC version numbers. It may be `C++11`-support related.

CC @dgquintas, @jtattermusch

--- a/Makefile
+++ b/Makefile
@@ -1123,11 +1123,7 @@ third_party/protobuf/configure:
 
 $(LIBDIR)/$(CONFIG)/protobuf/libprotobuf.a: third_party/protobuf/configure
 	$(E) "[MAKE]    Building protobuf"
-ifeq ($(HAVE_CXX11),true)
-	$(Q)(cd third_party/protobuf ; CC="$(CC)" CXX="$(CXX)" LDFLAGS="$(LDFLAGS_$(CONFIG)) -g" CXXFLAGS="-DLANG_CXX11 -std=c++11" CPPFLAGS="$(PIC_CPPFLAGS) $(CPPFLAGS_$(CONFIG)) -g" ./configure --disable-shared --enable-static)
-else
-	$(Q)(cd third_party/protobuf ; CC="$(CC)" CXX="$(CXX)" LDFLAGS="$(LDFLAGS_$(CONFIG)) -g" CXXFLAGS="-std=c++0x" CPPFLAGS="$(PIC_CPPFLAGS) $(CPPFLAGS_$(CONFIG)) -g" ./configure --disable-shared --enable-static)
-endif
+	$(Q)(cd third_party/protobuf ; CC="$(CC)" CXX="$(CXX)" LDFLAGS="$(LDFLAGS_$(CONFIG)) -g" CPPFLAGS="$(PIC_CPPFLAGS) $(CPPFLAGS_$(CONFIG)) -g" ./configure --disable-shared --enable-static)
 	$(Q)$(MAKE) -C third_party/protobuf clean
 	$(Q)$(MAKE) -C third_party/protobuf
 	$(Q)mkdir -p $(LIBDIR)/$(CONFIG)/protobuf

--- a/templates/Makefile.template
+++ b/templates/Makefile.template
@@ -673,11 +673,7 @@ third_party/protobuf/configure:
 
 $(LIBDIR)/$(CONFIG)/protobuf/libprotobuf.a: third_party/protobuf/configure
 	$(E) "[MAKE]    Building protobuf"
-ifeq ($(HAVE_CXX11),true)
-	$(Q)(cd third_party/protobuf ; CC="$(CC)" CXX="$(CXX)" LDFLAGS="$(LDFLAGS_$(CONFIG)) -g" CXXFLAGS="-DLANG_CXX11 -std=c++11" CPPFLAGS="$(PIC_CPPFLAGS) $(CPPFLAGS_$(CONFIG)) -g" ./configure --disable-shared --enable-static)
-else
-	$(Q)(cd third_party/protobuf ; CC="$(CC)" CXX="$(CXX)" LDFLAGS="$(LDFLAGS_$(CONFIG)) -g" CXXFLAGS="-std=c++0x" CPPFLAGS="$(PIC_CPPFLAGS) $(CPPFLAGS_$(CONFIG)) -g" ./configure --disable-shared --enable-static)
-endif
+	$(Q)(cd third_party/protobuf ; CC="$(CC)" CXX="$(CXX)" LDFLAGS="$(LDFLAGS_$(CONFIG)) -g" CPPFLAGS="$(PIC_CPPFLAGS) $(CPPFLAGS_$(CONFIG)) -g" ./configure --disable-shared --enable-static)
 	$(Q)$(MAKE) -C third_party/protobuf clean
 	$(Q)$(MAKE) -C third_party/protobuf
 	$(Q)mkdir -p $(LIBDIR)/$(CONFIG)/protobuf

